crates/ide_assists/src/handlers/generate_function.rs

```diff
@@ -1195,6 +1382,159 @@ fn foo() {
 async fn bar(arg: i32) ${0:-> ()} {
     todo!()
 }
+",
+        )
+    }
+
+    #[test]
+    fn create_method() {
+        check_assist(
+            generate_method,
+            r"
+struct S;
+
+fn foo() {
+    S.bar$0();
+}
+
+",
+            r"
+struct S;
+
+fn foo() {
+    S.bar();
+}
+impl S {
+
+
+fn bar(&self) ${0:-> ()} {
+    todo!()
+}
+}
+
+",
+        )
+    }
+
+    #[test]
+    fn create_method_within_an_impl() {
+        check_assist(
+            generate_method,
+            r"
+struct S;
+
+fn foo() {
+    S.bar$0();
+}
+impl S {}
+
+",
+            r"
+struct S;
+
+fn foo() {
+    S.bar();
+}
+impl S {
+    fn bar(&self) ${0:-> ()} {
+        todo!()
+    }
+}
+
+",
```

Comment:
as per style guide, we try to minize the percieved size of the test data:

https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#minimal-tests

Here'd, i'd written this as 

```
            r"
struct S;
fn foo() { S.bar$0(); }
impl S {}
",
```

(no extra blank line, function body on the same line)